### PR TITLE
Add notions of evaluated kinds in TheoryModel

### LINKED ISSUE #1118: Clean up special cases for LAMBDA and BITVECTOR_ACKERMANIZE in TheoryModel::getValue
The TheoryModel::getValue has a number of special cases which are hacks to make things work.

This is related to issue #970.

### PR DESCRIPTION
**Pull request by ajreynol**
This removes many of the special cases of getModelValue in TheoryModel based on a new notion of a kind being "evaluated", "unevaluated" or "semi-evaluated" in the model.

It also makes the status of APPLY_UF configurable in the above sense. In the case that we are interested in models but not in function values, we can configure APPLY_UF to be "semi-evaluated". This is of interest to sygus solving where we are not interesting in building model values for "evaluation functions".

This fixes #1118 and eliminates a 20% overhead from sygus solving.

Marking as complex because it would be good to have two reviews.

### CODE REVIEW

**Comment by ajreynol** [comment]
@barrettcw you may be interested in this since it involves TheoryModel.

File: src/theory/theory_model.cpp [hunk 1]
```diff
@@ -477,6 +482,17 @@ void TheoryModel::recordApproximation(TNode n, TNode pred)
...
 
+void TheoryModel::setUnevaluatedKind(Kind k)
+{
+  d_not_evaluated_kinds.insert(k);
+}
+
+void TheoryModel::setSemiEvaluatedKind(Kind k)
+{
+  d_not_evaluated_kinds.insert(k);
+  d_semi_evaluated_kinds.insert(k);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/theory_model.cpp]: ```d_semi_evaluated``` is used here: https://github.com/CVC4/CVC4/pull/1947/files/6452e8f5fd36603384935dc385a71e6ec02482df#diff-c473e454b74828446a5f7030f7840815R255

File: src/theory/theory_model.cpp [hunk 2]
```diff
@@ -160,148 +166,155 @@ Node TheoryModel::getModelValue(TNode n, bool hasBoundVars, bool useDontCares) c
...
-      Debug("model-getvalue-debug") << "ret (post-rewrite): " << ret << std::endl;
-      if(ret.getKind() == kind::CARDINALITY_CONSTRAINT) {
-        Debug("model-getvalue-debug") << "get cardinality constraint " << ret[0].getType() << std::endl;
-        ret = NodeManager::currentNM()->mkConst(getCardinality(ret[0].getType().toType()).getFiniteCardinality() <= ret[1].getConst<Rational>().getNumerator());
-      }else if(ret.getKind() == kind::CARDINALITY_VALUE) {
-        Debug("model-getvalue-debug") << "get cardinality value " << ret[0].getType() << std::endl;
-        ret = NodeManager::currentNM()->mkConst(Rational(getCardinality(ret[0].getType().toType()).getFiniteCardinality()));
+      else
+      {
+        // TODO: if func models not enabled, throw an error?
```
Comment (aniemetz) [inline src/theory/theory_model.cpp]: Can you maybe get rid of this TODO?

File: src/smt/smt_engine.cpp [hunk 3]
```diff
@@ -1313,6 +1313,11 @@ void SmtEngine::setDefaults() {
     {
       options::cbqiMidpoint.set(true);
     }
+    // do not assign function values (optimization)
+    if (!options::assignFunctionValues.wasSetByUser())
+    {
+      options::assignFunctionValues.set(false);
+    }
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/smt/smt_engine.cpp]: This is the part of the code where options are set depending upon the input format, where we want to disable this option only if the input is sygus.  This is the earliest we can disable this option I believe.

File: src/theory/theory_model.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/theory_model.cpp]: Yes, I think this TODO can go away.

**Comment by ajreynol** [comment]
@aniemetz Regarding #1901, my goal in this PR was not to change the behavior of the model, just reorganize it.  As expected, this PR does not impact it, I disabled the option restriction and it still fails --check-model. Although my hope is that this will make it easier to debug that issue (where now you have control over how to configure the model for the BV_ACKERMANNIZE kinds).

**Pull request merged**